assets/javascripts/discourse/components/babble-composer.js.es6

```diff
@@ -73,6 +77,144 @@ export default Ember.Component.extend({
+      this._addText(this._getSelected(), placeholder);
+
+      if (data.xhr && data.originalFiles.length === 1) {
+        this.set("isCancellable", true);
+        this._xhr = data.xhr();
+      }
+    });
+
+    $element.on("fileuploadfail", (e, data) => {
+      this._resetUpload(true);
+
+      const userCancelled = this._xhr && this._xhr._userCancelled;
+      this._xhr = null;
+
+      if (!userCancelled) {
+        Discourse.Utilities.displayErrorForUpload(data);
+      }
+    });
+
+    this.messageBus.subscribe("/uploads/composer", upload => {
+      // replace upload placeholder
+      if (upload && upload.url) {
+        if (!this._xhr || !this._xhr._userCancelled) {
+          const markdown = Discourse.Utilities.getUploadMarkdown(upload);
+          this.set('text', this.get('text').replace(uploadPlaceholder, markdown));
+          this._resetUpload(false);
+          this.set('showUpload', false)
+        } else {
+          this._resetUpload(true);
+        }
+      } else {
+        this._resetUpload(true);
+        Discourse.Utilities.displayErrorForUpload(upload);
+      }
+    });
+  },
+
+  _resetUpload: function(removePlaceholder) {
+    this._validUploads--;
+    if (this._validUploads === 0) {
+      this.setProperties({ uploadProgress: 0, isUploading: false, isCancellable: false });
+    }
+    if (removePlaceholder) {
+      this.set('text', this.get('text').replace(this.get('placeholder'), ""));
+    }
+  },
+
+  _unbindUploadTarget: function() {
+    this._validUploads = 0;
+    this.messageBus.unsubscribe("/uploads/composer");
+    const $uploadTarget = this.$();
+    try { $uploadTarget.fileupload("destroy"); }
+    catch (e) { }
+    $uploadTarget.off();
+  }.on('willDestroyElement'),
+
+  _addTextFromMenu: function() {
+    this._addText(this._getSelected(), this.get('addText'))
+  }.observes('addText'),
+
+  _addText: function(sel, text) {
+    const insert = `${sel.pre}${text}`;
+    this.set('text', `${insert}${sel.post}`);
+    this._selectText(insert.length, 0);
+    Ember.run.scheduleOnce("afterRender", () => this.$("textarea").focus());
+  },
+
+  _selectText: function(from, length) {
+    Ember.run.scheduleOnce('afterRender', () => {
+      const $textarea = this.$('textarea');
+      const textarea = $textarea[0];
+      const oldScrollPos = $textarea.scrollTop();
+      if (!this.capabilities.isIOS) {
+        $textarea.focus();
+      }
+      textarea.selectionStart = from;
+      textarea.selectionEnd = textarea.selectionStart + length;
+      $textarea.scrollTop(oldScrollPos);
+    });
+  },
+
+  _getSelected: function(trimLeading) {
+    const textarea = this.$('textarea')[0];
+    const value = textarea.value;
+    var start = textarea.selectionStart;
+    let end = textarea.selectionEnd;
+
+    // trim trailing spaces cause **test ** would be invalid
+    while (end > start && /\s/.test(value.charAt(end-1))) {
```

Comment:
Whoa, do we need to do it this way? Whatever happened to good ol' `"  string  ".trim()`?